doc/man/tapyrus-genesis.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
.TH TAPYRUS-GENESIS "1" "March 2026" "tapyrus-genesis v0.7.2" "User Commands"
.SH NAME
tapyrus-genesis \- manual page for tapyrus-genesis v0.7.2
.SH SYNOPSIS
.B tapyrus-genesis
[\fI\,options\/\fR]
.SH DESCRIPTION
Tapyrus Core tapyrus\-genesis utility version v0.7.2
.IP
Create hex\-encoded tapyrus genesis block
.SH OPTIONS
.HP
\-?
.IP
This help message
.PP
Chain selection options:
.PP
Genesis block:
.HP
\fB\-address=\fR<pay_to_address>
.IP
Optional. Specify coinbase script pay to address.
.HP
\fB\-dev\fR
.IP
Specify dev environment.
.HP
\fB\-generatekey\fR
.IP
Generate a public key, private key pair in dev mode.
.HP
\fB\-signblockprivatekey=\fR<privatekey\-WIF>
.IP
Optional. Sets the aggregate private key in WIF to be used to sign
genesis block. If it is not set, this command creates no proof in
genesis block.
.HP
\fB\-signblockpubkey=\fR<pubkey>
.IP
Sets the aggregate public key for Signed Blocks
.HP
\fB\-time=\fR<time>
.IP
Specify genesis block time as UNIX Time. If this don't set, use current
time.
.SH COPYRIGHT
Copyright (C) 2009-2019 The Bitcoin Core developers
Copyright (C) 2019-2026 Chaintope Inc.

Please contribute if you find Tapyrus Core useful. Visit
<https://www.chaintope.com/en/chaintope-blockchain-protocol> for further
information about the software.
The source code is available from <https://github.com/chaintope/tapyrus-core>.

This is experimental software.
Distributed under the MIT software license, see the accompanying file COPYING
or <https://opensource.org/licenses/MIT>


doc/release-notes.md
(note: this is a temporary file, to be added-to by anybody, and moved to
release-notes at release time)

Bitcoin Core version 0.17.0 is now available from:

  <https://bitcoincore.org/bin/bitcoin-core-0.17.0/>

This is a new major version release, including new features, various bugfixes
and performance improvements, as well as updated translations.

Please report bugs using the issue tracker at GitHub:

  <https://github.com/bitcoin/bitcoin/issues>

To receive security and update notifications, please subscribe to:

  <https://bitcoincore.org/en/list/announcements/join/>

How to Upgrade
==============

If you are running an older version, shut it down. Wait until it has completely
shut down (which might take a few minutes for older versions), then run the
installer (on Windows) or just copy over `/Applications/Bitcoin-Qt` (on Mac)
or `bitcoind`/`bitcoin-qt` (on Linux).

If your node has a txindex, the txindex db will be migrated the first time you run 0.17.0 or newer, which may take up to a few hours. Your node will not be functional until this migration completes.

The first time you run version 0.15.0 or newer, your chainstate database will be converted to a
new format, which will take anywhere from a few minutes to half an hour,
depending on the speed of your machine.

Note that the block database format also changed in version 0.8.0 and there is no
automatic upgrade code from before version 0.8 to version 0.15.0. Upgrading
directly from 0.7.x and earlier without redownloading the blockchain is not supported.
However, as usual, old wallet versions are still supported.

Downgrading warning
-------------------

The chainstate database for this release is not compatible with previous
releases, so if you run 0.15 and then decide to switch back to any
older version, you will need to run the old release with the `-reindex-chainstate`
option to rebuild the chainstate data structures in the old format.

If your node has pruning enabled, this will entail re-downloading and
processing the entire blockchain.

Compatibility
==============

Bitcoin Core is extensively tested on multiple operating systems using
the Linux kernel, macOS 10.10+, and Windows 7 and newer (Windows XP is not supported).

Bitcoin Core should also work on most other Unix-like systems but is not
frequently tested on them.

From 0.17.0 onwards macOS <10.10 is no longer supported. 0.17.0 is built using Qt 5.9.x, which doesn't
support versions of macOS older than 10.10.

Known issues
============

- Upgrading from 0.13.0 or older currently results in memory blow-up during the roll-back of blocks to the SegWit activation point. In these cases, a full `-reindex` is necessary.

- The GUI suffers from visual glitches in the new MacOS dark mode. This has to do with our Qt theme handling and is not a new problem in 0.17.0, but is expected to be resolved in 0.17.1.

Notable changes
===============

Changed configuration options
-----------------------------

- `-includeconf=<file>` can be used to include additional configuration files.
  Only works inside the `bitcoin.conf` file, not inside included files or from
  command-line. Multiple files may be included. Can be disabled from command-
  line via `-noincludeconf`. Note that multi-argument commands like
  `-includeconf` will override preceding `-noincludeconf`, i.e.
  ```
  noincludeconf=1
  includeconf=relative.conf
  ```

  as bitcoin.conf will still include `relative.conf`.

GUI changes
-----------

- Block storage can be limited under Preferences, in the Main tab. Undoing this setting requires downloading the full blockchain again. This mode is incompatible with -txindex and -rescan.

External wallet files
---------------------

The `-wallet=<path>` option now accepts full paths instead of requiring wallets
to be located in the -walletdir directory.

Newly created wallet format
---------------------------

If `-wallet=<path>` is specified with a path that does not exist, it will now
create a wallet directory at the specified location (containing a wallet.dat
data file, a db.log file, and database/log.?????????? files) instead of just
creating a data file at the path and storing log files in the parent
directory. This should make backing up wallets more straightforward than
before because the specified wallet path can just be directly archived without
having to look in the parent directory for transaction log files.

For backwards compatibility, wallet paths that are names of existing data files
in the `-walletdir` directory will continue to be accepted and interpreted the
same as before.

Dynamic loading and creation of wallets
---------------------------------------

Previously, wallets could only be loaded or created at startup, by specifying `-wallet` parameters on the command line or in the bitcoin.conf file. It is now possible to load, create and unload wallets dynamically at runtime:

- Existing wallets can be loaded by calling the `loadwallet` RPC. The wallet can be specified as file/directory basename (which must be located in the `walletdir` directory), or as an absolute path to a file/directory.
- New wallets can be created (and loaded) by calling the `createwallet` RPC. The provided name must not match a wallet file in the `walletdir` directory or the name of a wallet that is currently loaded.
- Loaded wallets can be unloaded by calling the `unloadwallet` RPC.

This feature is currently only available through the RPC interface.

Coin selection
--------------

### Partial spend avoidance

When an address is paid multiple times the coins from those separate payments can be spent separately which hurts privacy due to linking otherwise separate addresses. A new `-avoidpartialspends` flag has been added (default=false). If enabled, the wallet will always spend existing UTXO to the same address together even if it results in higher fees. If someone were to send coins to an address after it was used, those coins will still be included in future coin selections.


src/tapyrus-genesis.cpp
// Copyright (c) 2019-2021 Chaintope Inc.
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.
#include <tapyrus-config.h>

#include <clientversion.h>
#include <util.h>
#include <federationparams.h>
#include <chainparams.h>
#include <utilstrencodings.h>
#include <key_io.h>
#include <streams.h>
#include <consensus/validation.h>
#include <validation.h>
#include <tinyformat.h>
#include <script/standard.h>
#include <stdio.h>
#include <xfieldhistory.h>

static const int CONTINUE_EXECUTION=-1;

static void SetupTapyrusGenesisArgs()
{
    gArgs.AddArg("-?", "This help message", false, OptionsCategory::OPTIONS);
    SetupFederationParamsOptions();

    // Signed Blocks options
    gArgs.AddArg("-signblockpubkey=<pubkey>", "Sets the aggregate public key for Signed Blocks", false, OptionsCategory::GENESIS);
    gArgs.AddArg("-signblockprivatekey=<privatekey-WIF>", "Optional. Sets the aggregate private key in WIF to be used to sign genesis block. If it is not set, this command creates no proof in genesis block.", false, OptionsCategory::GENESIS);

    // Genesis Block options
    gArgs.AddArg("-time=<time>", "Specify genesis block time as UNIX Time. If this don't set, use current time.", false, OptionsCategory::GENESIS);
    gArgs.AddArg("-address=<pay_to_address>", "Optional. Specify coinbase script pay to address.", false, OptionsCategory::GENESIS);

    // Dev mode option
    gArgs.AddArg("-dev", "Specify dev environment.", false, OptionsCategory::GENESIS);
    gArgs.AddArg("-generatekey", "Generate a public key, private key pair in dev mode.", false, OptionsCategory::GENESIS);

    // Hidden
    gArgs.AddArg("-h", "", false, OptionsCategory::HIDDEN);
    gArgs.AddArg("-help", "", false, OptionsCategory::HIDDEN);
}

//
// This function returns either one of EXIT_ codes when it's expected to stop the process or
// CONTINUE_EXECUTION when it's expected to continue further.
//
static int AppInit(int argc, char* argv[])
{
    //
    // Parameters
    //
    SetupTapyrusGenesisArgs();
    std::string error;
    if (!gArgs.ParseParameters(argc, argv, error)) {
        fprintf(stderr, "Error parsing command line arguments: %s\n", error.c_str());
        return EXIT_FAILURE;
    }

    if (argc < 2 || HelpRequested(gArgs)) {
        // First part of help message is specific to this utility
        std::string strUsage = PACKAGE_NAME " tapyrus-genesis utility version " + FormatFullVersion() + "\n\n" +
                               "Usage:   tapyrus-genesis [options]\n" +
                               "         Create hex-encoded tapyrus genesis block\n" +
                               "\n";
        strUsage += gArgs.GetHelpMessage();

        fprintf(stdout, "%s", strUsage.c_str());

        if (argc < 2) {
            fprintf(stderr, "Error: too few parameters\n");
            return EXIT_FAILURE;
        }
        return EXIT_SUCCESS;
    }

    // Check for -dev parameter (Params() calls are only valid after this clause)
    try {
        SelectParams(gArgs.GetChainMode());
        SelectFederationParams(gArgs.GetChainMode(), false);
    } catch (const std::exception& e) {
        fprintf(stderr, "Error: %s\n", e.what());
        return EXIT_FAILURE;
    }

    return CONTINUE_EXECUTION;
}

static int generateNewKeyPair()
{
    //generate a secret key
    CKey secret;
    secret.MakeNewKey(true);//compressed

    //generate its public key
    CPubKey pubkey = secret.GetPubKey();
    assert(pubkey.IsFullyValid());
    assert(secret.VerifyPubKey(pubkey));

    fprintf(stdout, "private key: %s\n", EncodeSecret(secret).c_str());
    fprintf(stdout, "public key: %s\n", HexStr(pubkey.begin(), pubkey.end()).c_str());

    return EXIT_SUCCESS;
}

static int generateGenesis(CPubKey aggpubkey, CKey& privatekey, long long blockTime, std::string& payToAddress)
{
    CBlock genesis { createGenesisBlock(aggpubkey, privatekey, blockTime, payToAddress) };

    //initialize xfield history
    CXFieldHistory history(genesis);

    // check validity
    CValidationState state;
    bool fCheckProof = privatekey.IsValid();
    if(!CheckBlock(genesis, state, fCheckProof)) {
        fprintf(stderr, "error: Consensus::CheckBlock: %s", FormatStateMessage(state).c_str());
        return EXIT_FAILURE;
    }

    // serialize
    CDataStream ss(SER_NETWORK, PROTOCOL_VERSION);
    ss << genesis;
    std::string genesis_hex = HexStr(ss.begin(), ss.end());

    // print
    fprintf(stdout, "%s\n", genesis_hex.c_str());
    return EXIT_SUCCESS;
}

static int CommandLine()
{
    auto generateKey = gArgs.GetBoolArg("-generatekey", false);
    if(generateKey && gArgs.GetChainMode() != TAPYRUS_OP_MODE::DEV)
    {
        fprintf(stderr, "Error: generateKey is supported only in DEV mode.\n");
        return EXIT_FAILURE;
    }

    CPubKey aggpubkey;
    try {
        std::string pubkey = gArgs.GetArg("-signblockpubkey", "");
        if(pubkey.size())
        {
            std::vector<unsigned char> key( ParseHex(pubkey) );
            aggpubkey = CPubKey(key.begin(), key.end());
            if(!aggpubkey.IsFullyValid())
            {
                fprintf(stderr, "Error: Aggregate Public Key was invalid.\n");
                return EXIT_FAILURE;
            }
            if(!aggpubkey.IsCompressed())
            {
                fprintf(stderr, "Error: Uncompressed Aggregate Public Key is not supported.\n");
                return EXIT_FAILURE;
            }
        }
    } catch (const std::exception& e) {
        fprintf(stderr, "Error: %s\n", e.what());
        return EXIT_FAILURE;
    }

    // This is for using CKey.sign().
    ECC_Start();

    // This is for using CPubKey.verify().
    ECCVerifyHandle globalVerifyHandle;

    std::string wif = gArgs.GetArg("-signblockprivatekey", "");
    CKey privatekey(DecodeSecret(wif));
    if(wif.size() && !privatekey.IsValid())
    {
        fprintf(stderr, "Error: Aggregate private key was invalid.\n");
        return EXIT_FAILURE;
    }

    if(privatekey.IsValid() && aggpubkey != privatekey.GetPubKey())
    {
        fprintf(stderr, "Error: Aggregate private key does not correspond to given Aggregate public key.\n");
        return EXIT_FAILURE;
    }

    auto blockTime = gArgs.GetArg("-time", 0);
    if(!blockTime) {
        blockTime = time(0);
    }

    auto payToAddress = gArgs.GetArg("-address", "");
     if(payToAddress.size() && !IsValidDestinationString(payToAddress)) {
         fprintf(stderr, "Error: Invalid address specified in -address option.\n");
         return EXIT_FAILURE;
     }

    //generate after ECC initialization
    if(generateKey)
        return generateNewKeyPair();
    else
        return generateGenesis(aggpubkey, privatekey, blockTime, payToAddress);
}

int main(int argc, char* argv[])
{
    SetupEnvironment();

    try {
        int ret = AppInit(argc, argv);
        if (ret != CONTINUE_EXECUTION)
            return ret;
    }
    catch (const std::exception& e) {
        PrintExceptionContinue(&e, "AppInit()");
        return EXIT_FAILURE;
    } catch (...) {
        PrintExceptionContinue(nullptr, "AppInit()");
        return EXIT_FAILURE;
    }

    int ret = EXIT_FAILURE;
    try {
        ret = CommandLine();
    }
    catch (const std::exception& e) {
        PrintExceptionContinue(&e, "CommandLine()");
    } catch (...) {
        PrintExceptionContinue(nullptr, "CommandLine()");
    }
    return ret;
}